main.go

```diff
@@ -39,65 +40,78 @@ func main() {
 	go runEthListener(ethEvents, &currentBlock)
 	go terminateExpiringOrgs(stateEvents, &currentBlock)
 
+	channels := map[string]OrgChannel{}
+	channelCloseEvents := make(chan string)
+
 	for {
+		select {
 		// stream events from contract
-		e := <-ethEvents
-
-		if err := db.UpsertEvent(e); err != nil {
-			l.Fatal("Failed to upsert event", e, err)
-		}
+		case e := <-ethEvents:
+			if err := db.UpsertEvent(e); err != nil {
+				l.Fatal("Failed to upsert event", e, err)
+			}
 
-		// if event has been removed, a reorg has happened
-		if e.Removed {
-			stateAfterRemoval(&e, &currentBlock)
-		}
+			// if event has been removed, a reorg has happened
+			if e.Removed {
+				stateAfterRemoval(&e, &currentBlock)
+			}
 
-		if queue, ok := queues.LoadOrStore(e.Org, utils.NewEventQueue(e)); ok {
-			// queue already exists, and a goroutine is running, just add event
-			q := queue.(*utils.EventQueue)
-			q.AddEvent(e)
-			// between LoadOrStore and AddEvent, queue might get deleted from map
-			if _, ok := queues.LoadOrStore(e.Org, q); !ok {
-				// re-spawn another goroutine for this org
-				go processEventsForOrg(e.Org, stateEvents)
+			ch := getOrCreateOrgChannel(e.Org, &channels)
+			ch.event <- e
+			select {
+			case <-ch.ack:
+			case <-time.After(3 * time.Second):
+				// if we didn't get the ack, we assume the processor is not running
+				go processEventsForOrg(e.Org, ch, stateEvents, channelCloseEvents)
+				<-ch.ack
 			}
-		} else {
-			// spawn a goroutine to process newly created queue
-			go processEventsForOrg(e.Org, stateEvents)
+		// processor for an org has cleared queue
+		case closedOrg := <-channelCloseEvents:
+			delete(channels, closedOrg)
 		}
 	}
 }
 
-func processEventsForOrg(org string, stateEvents chan db.Dep) {
-	queue, ok := queues.Load(org)
-	if !ok {
-		l.Fatal("Queue was missing for org", org)
+func getOrCreateOrgChannel(org string, channels *map[string]OrgChannel) OrgChannel {
+	chs := *channels
```

Comment:
Why do we need to dereference? Haven't written Go in years...

Reply:
Because we use indexing and compiler is not smart enough to do it for us, if we don't dereference we'd get:

```
invalid operation: cannot index channels (variable of type *map[string]OrgChannel)compilerNonIndexableOperand)
```